File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamfriends/callback/ChatMsgCallback.java
package in.dragonbra.javasteam.steam.handlers.steamfriends.callback;

import in.dragonbra.javasteam.enums.EChatEntryType;
import in.dragonbra.javasteam.generated.MsgClientChatMsg;
import in.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg;
import in.dragonbra.javasteam.types.SteamID;

import java.nio.charset.StandardCharsets;

/**
 * This callback is fired when a chat room message arrives.
 */
public class ChatMsgCallback extends CallbackMsg {

    private SteamID chatterID;

    private SteamID chatRoomID;

    private EChatEntryType chatMsgType;

    private String message;

    public ChatMsgCallback(MsgClientChatMsg msg, byte[] payload) {
        chatterID = msg.getSteamIdChatter();
        chatRoomID = msg.getSteamIdChatRoom();
        chatMsgType = msg.getChatMsgType();

        message = new String(payload, StandardCharsets.UTF_8).replaceAll("\0+$", ""); // trim any extra null chars from the end
    }

    /**
     * @return the {@link SteamID} of the chatter.
     */
    public SteamID getChatterID() {
        return chatterID;
    }

    /**
     * @return the {@link SteamID} of the chat room.
     */
    public SteamID getChatRoomID() {
        return chatRoomID;
    }

    /**
     * @return chat entry type.
     */
    public EChatEntryType getChatMsgType() {
        return chatMsgType;
    }

    /**
     * @return the message.
     */
    public String getMessage() {
        return message;
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamfriends/callback/ChatMsgCallback.kt
package `in`.dragonbra.javasteam.steam.handlers.steamfriends.callback

import `in`.dragonbra.javasteam.base.ClientMsg
import `in`.dragonbra.javasteam.base.IPacketMsg
import `in`.dragonbra.javasteam.enums.EChatEntryType
import `in`.dragonbra.javasteam.generated.MsgClientChatMsg
import `in`.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg
import `in`.dragonbra.javasteam.types.SteamID
import java.nio.charset.StandardCharsets

/**
 * This callback is fired when a chat room message arrives.
 */
@Suppress("MemberVisibilityCanBePrivate")
class ChatMsgCallback(packetMsg: IPacketMsg) : CallbackMsg() {

    /**
     * Gets the [SteamID] of the chatter.
     */
    val chatterID: SteamID

    /**
     * Gets the [SteamID] of the chat room.
     */
    val chatRoomID: SteamID

    /**
     * Gets chat entry type.
     */
    val chatMsgType: EChatEntryType

    /**
     * Gets the message.
     */
    val message: String

    init {
        val chatMsg = ClientMsg(MsgClientChatMsg::class.java, packetMsg)
        val msg = chatMsg.body

        chatterID = msg.steamIdChatter
        chatRoomID = msg.steamIdChatRoom

        chatMsgType = msg.chatMsgType

        // trim any extra null chars from the end
        val payload = chatMsg.payload
        message = String(payload.toByteArray(), StandardCharsets.UTF_8).replace("\u0000+$".toRegex(), "")
    }
}
